extensions/guacamole-auth-radius/src/main/java/org/apache/guacamole/auth/radius/RadiusConnectionService.java

```diff
@@ -0,0 +1,394 @@
+        if (radAuth instanceof PEAPAuthenticator || 
+            radAuth instanceof EAPTLSAuthenticator || 
+            radAuth instanceof EAPTTLSAuthenticator) {
+
+            if (caFile != null && !caFile.isEmpty())
+                ((EAPTLSAuthenticator)radAuth).setCaFile(basePath + caFile);
+
+            if (caType != null && !caType.isEmpty())
+                ((EAPTLSAuthenticator)radAuth).setCaFileType(caType);
+
+            if (caPassword != null && !caPassword.isEmpty())
+                ((EAPTLSAuthenticator)radAuth).setCaPassword(caPassword);
+
+            if (keyFile != null && !keyFile.isEmpty())
+                ((EAPTLSAuthenticator)radAuth).setKeyFile(basePath + keyFile);
+
+            if (keyType != null && !keyType.isEmpty())
+                ((EAPTLSAuthenticator)radAuth).setKeyFileType(keyType);
+
+            if (keyPassword != null && !keyPassword.isEmpty())
+                ((EAPTLSAuthenticator)radAuth).setKeyPassword(keyPassword);
+
+            ((EAPTLSAuthenticator)radAuth).setTrustAll(trustAll);
+
+        }
+
+        // If we're using EAP-TTLS, we need to define tunneled protocol
+        if (radAuth instanceof EAPTTLSAuthenticator) {
+
+            if (innerProtocol == null || innerProtocol.isEmpty())
+                return null;
+
+            ((EAPTTLSAuthenticator)radAuth).setInnerProtocol(innerProtocol);
+
+        }
+
+        return radAuth;
+
+    }
+
+    /**
+     * Authenticate to the RADIUS server and return the response from the
+     * server.
+     *
+     * @param username
+     *     The username for authentication.
+     * @param password
+     *     The password for authentication.
+     *
+     * @return
+     *     A RadiusPacket with the response of the server.
+     *
+     * @throws GuacamoleException
+     *     If an error occurs while talking to the server. 
+     */
+    public RadiusPacket authenticate(String username, String password) 
+            throws GuacamoleException {
+
+        // Create the connection and load the attribute dictionary
+	createRadiusConnection();
+        AttributeFactory.loadAttributeDictionary("net.jradius.dictionary.AttributeDictionaryImpl");
+
+        // If the client is null, we return null - something has gone wrong
+        if (radiusClient == null)
+            return null;
+
+        // If a username hasn't been provided, stop
+        if (username == null || username.isEmpty()) {
+            logger.warn("Anonymous access not allowed with RADIUS client.");
+            return null;
+        }
+
+        // If a password hasn't been provided, stop
+        if (password == null || password.isEmpty()) {
+            logger.warn("Password required for RADIUS authentication.");
+            return null;
+        }
+
+        RadiusAuthenticator radAuth = setupRadiusAuthenticator();
+
+        if (radAuth == null)
+            throw new GuacamoleException("Unknown RADIUS authentication protocol.");
```

Comment:
`setupRadiusAuthenticator()` returns `null` for many cases, not simply that the RADIUS authentication protocol is unknown. Probably best that `setupRadiusAuthenticator()` throw proper exceptions when it encounters error conditions